### PR TITLE
evaluator: just one time eval

### PR DESCRIPTION
**Pull request by zxylvlp**

### CODE REVIEW

File: ast/flag.go [hunk 1]
```diff
@@ -17,7 +17,22 @@ const preEvaluable = FlagHasParamMarker | FlagHasFunc | FlagHasVariable | FlagHa
 
 // IsPreEvaluable checks if the expression can be evaluated before execution.
 func IsPreEvaluable(expr ExprNode) bool {
-	return expr.GetFlag()|preEvaluable == preEvaluable
+	return (expr.GetFlag()|preEvaluable)&^FlagEvaluated == preEvaluable
```
Comment (coocood) [inline ast/flag.go]: Add ` | FlagEvaluated` to `preEvaluable` should be better.

File: evaluator/evaluator.go [hunk 2]
```diff
@@ -38,21 +38,21 @@ const (
...
 func EvalDatum(ctx context.Context, expr ast.ExprNode) (d types.Datum, err error) {
+	if ast.IsEvaluated(expr) {
+		return *expr.GetDatum(), nil
+	}
 	e := &Evaluator{ctx: ctx}
 	expr.Accept(e)
 	if e.err != nil {
 		return d, errors.Trace(e.err)
 	}
+	expr.SetFlag(expr.GetFlag() | ast.FlagEvaluated)
```
Comment (coocood) [inline evaluator/evaluator.go]: If set `FlagEvaluated` only if it is PreEvaluable, then `IsEvaluated` can just return `flag&FlagEvaluated == FlagEvaluated`.

**Comment by coocood** [comment]
For prepared statement, when a statement has finished execution, all expressions should reset `FlagEvaluated`.

**Comment by shenli** [comment]
Any test case to ensure this PR works?

**Comment by zxylvlp** [comment]
@shenli  I think it just a optimization of performance, the the logical test cann't do it.

**Comment by zxylvlp** [comment]
@coocood @shenli @zimulala PTAL

File: evaluator/evaluator.go [hunk 3]
```diff
@@ -52,7 +52,9 @@ func EvalDatum(ctx context.Context, expr ast.ExprNode) (d types.Datum, err error
 	if e.err != nil {
 		return d, errors.Trace(e.err)
 	}
-	expr.SetFlag(expr.GetFlag() | ast.FlagEvaluated)
+	if ast.IsPreEvaluable(expr) && !ast.IsConstant(expr) && (expr.GetFlag()&ast.FlagHasFunc == 0) {
```
Comment (coocood) [inline evaluator/evaluator.go]: We don't need to check `IsConstatnt` and `FlagHasFunc`.

Comment (zxylvlp) [inline evaluator/evaluator.go]: When FlagHasFunc has a rand(), it can't be evaluated.

File: ast/flag.go [hunk 4]
```diff
@@ -13,13 +13,28 @@
...
 
-const preEvaluable = FlagHasParamMarker | FlagHasFunc | FlagHasVariable | FlagHasDefault
+const preEvaluable = FlagHasParamMarker | FlagHasFunc | FlagHasVariable | FlagHasDefault | FlagPreEvaluated
 
 // IsPreEvaluable checks if the expression can be evaluated before execution.
 func IsPreEvaluable(expr ExprNode) bool {
 	return expr.GetFlag()|preEvaluable == preEvaluable
 }
 
+// ResetEvaluatedFlag reset the evaluated flag.
```
Comment (zimulala) [inline ast/flag.go]: reset -> resets ?

File: ast/flag.go [hunk 5]
```diff
@@ -163,3 +192,13 @@ func (f *flagSetter) aggregateFunc(x *AggregateFuncExpr) {
 	}
 	x.SetFlag(flag)
 }
+
+// SetBinaryOperationExprFlag set flag to binary operation expression generated manually.
```
Comment (zimulala) [inline ast/flag.go]: set -> sets ?

**Comment by zxylvlp** [comment]
@coocood @shenli @zimulala PTAL

**Comment by coocood** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**